### ISSUE TITLE
Referencing other local JSON Schemas

### ISSUE THREAD

**Issue by marksto** [manifold-systems/manifold#63]
I have _strong_ doubts whether this is a bug or not, but it is definitely a _nice-to-have_ feature which is foreseen by the JSON Schema specification.

## What the specification states

The [Schema References With "$ref"](https://json-schema.org/latest/json-schema-core.html#rfc.section.8.3.2) section states the following:

> Schemas can be identified by any URI that has been given to them, including a JSON Pointer or their URI given directly by "$id". In all cases, dereferencing a "$ref" reference involves first resolving its value as a URI reference against the current base URI per [RFC 3986](https://tools.ietf.org/html/rfc3986).

As an example it provides the following simple case:

> When an implementation encounters the reference to `"other.json"`, it resolves this to `<http://example.net/other.json>`, which is not defined in this document. If a schema with that identifier has otherwise been supplied to the implementation, it can also be used automatically.

And this is how it works in the latest version of the **Manifold**, no offense here. 😄 We are able to reference the schema with the same base URI, i.e. on the same level in the filesystem's directory structure. It compiles and it runs smoothly.

At the same time, it is impossible to introduce any hierarchical structure on top of one's schemas, i.e. to use subdirectories and reference other schemas relatively in both directions: forth with `"sub/other.json"` and back with `"../other.json"`.

But there are these _"can be identified by any URI"_ and _"In all cases <...> resolving its value as a URI reference against the current base URI"_ parts that make me believe that such a scenario is also considered possible, yet omitted in the current **Manifold**'s implementation.

## What the specification doesn't state yet

What causes my doubts is an associated cite reference:

> What should implementations do when the referenced schema is not known? Are there circumstances in which automatic network dereferencing is allowed? A same origin policy? A user-configurable option?  In the case of an evolving API described by Hyper-Schema, it is expected that new schemas will be added to the system dynamically, so placing an absolute requirement of pre-loading schema documents is not feasible.

While it is _not feasible_ in the wild (a.k.a. World Wild Web), we are dealing with a special case of "static resources as Java Types" and it would seem that it doesn't matter much. However, there might be matters to discuss.

What everyone (including myself) definitely wants to be able to do is to work with such references without fuss in the **Manifold**'s "dynamic mode".

### Test 1
Checks an ability to reference the `"sub/other.json"` schema from the current one.

**Manifold** is unable to compile such a schema in the current implementation:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project manifold-json-subschema-test: Compilation failure: Compilation failure:
[ERROR] /abc/res/Test1.java:[39,11] cannot find symbol
[ERROR] symbol:   class SubTest1
[ERROR] location: interface abc.res.Test1
[ERROR] /abc/res/Test1.java:[44,27] cannot find symbol
[ERROR] symbol:   class SubTest1
[ERROR] location: interface abc.res.Test1
[ERROR] /abc/res/Test1.java:[70,32] cannot find symbol
[ERROR] symbol:   class SubTest1
[ERROR] location: class abc.res.Test1.Builder
[ERROR] /abc/res/Test1.java:[82,27] cannot find symbol
[ERROR] symbol:   class SubTest1
[ERROR] location: class abc.res.Test1.Copier 
```

It also happens that **Manifold**'s IntelliJ plugin is dealing fine (w/o `"No such file or directory"` error) with such a `"$ref"`.

<a href="https://ibb.co/vmJPG19"><img src="https://i.ibb.co/TRH8Scy/test-1.png" alt="test-1" border="0"></a>

### Test 2
Checks an ability to back-reference the `"../other.json"` schema from the current one.

⚠️ In case **Test 1** succeeds, this one is also expected to pass according to the [URI Relative Resolution](https://tools.ietf.org/html/rfc3986#section-5.2.4) rules.

Now it gives almost the same compilation error:
```
[ERROR] /abc/res/sub/SubTest2.java:[39,11] cannot find symbol
[ERROR] symbol:   class Test2
[ERROR] location: interface abc.res.sub.SubTest2
[ERROR] /abc/res/sub/SubTest2.java:[44,25] cannot find symbol
[ERROR] symbol:   class Test2
[ERROR] location: interface abc.res.sub.SubTest2
[ERROR] /abc/res/sub/SubTest2.java:[70,30] cannot find symbol
[ERROR] symbol:   class Test2
[ERROR] location: class abc.res.sub.SubTest2.Builder
[ERROR] /abc/res/sub/SubTest2.java:[82,25] cannot find symbol
[ERROR] symbol:   class Test2
[ERROR] location: class abc.res.sub.SubTest2.Copier
```

And is equally located by the IntelliJ plugin.

<a href="https://ibb.co/gSz7zHs"><img src="https://i.ibb.co/9bsvszR/test-2.png" alt="test-2" border="0"></a>

---

As always, test project is available here:
https://github.com/marksto/manifold-json-subschema-test

**Comment by rsmckinney** [comment]
This is a relatively easy fix btw.  Let me now if there is time pressure for this, I can fix it sooner.

**Comment by marksto** [comment]
@rsmckinney, nope, there's no hurry. Hope that's really that easy as pluggin' in java.net.URL, doh!

**Comment by rsmckinney** [comment]
In progress

**Comment by rsmckinney** [comment]
Fix available in release 0.65-alpha

**Comment by marksto** [comment]
Superb, will check this out soon!
